### ISSUE TITLE
Need XrdSecEntity information to save checksum

### ISSUE THREAD

**Issue by bbockelm** [xrootd/xrootd#1294]
In the `xrootd-multiuser` library, we need information from the current `XrdSecEntity` object in order to save the checksum (as the credential will let us know the relevant UID to utilize -- and we need the  right "user" UID, not the system one).

However, it doesn't seem this is available to the `ofs.xattrlib` plugin.  Is there a thread-local where the credential is available?  I see that it's passed to the `XrdOfs` layer but not beyond that to the `XrdCks` layer.

I would prefer to not construct an entire parallel xattr universe as we had to do with HDFS.

**Comment by abh3** [comment]
I am somewhat confused. Are we talking about the extended attribute plugin, the checksum plugin, or both?

**Comment by bbockelm** [comment]
I suppose I only need this information in the extended attribute plugin in order to talk to the filesystem with the right credentials; if the credential is stored as a thread-local, I suppose the checksum plugin does not need this information.

There are obvious drawbacks to hiding this in thread state, however.  If we wanted it explicitly passed around, I assume the extended attribute plugin would have to get this information from the checksum plugin.

**Comment by abh3** [comment]
OK, it still seems somewhat odd. The extended attributes are always targeted to an existing file. Presumably, that file was created by the person asking for the checksum (or at least authorized to do so). If all you need is the uid that needs to be used, why not get it from the exiting file using stat()? Obviously, this is too simple so I must be missing the problem at hand..

**Comment by bbockelm** [comment]
Two thoughts come to mind:

1.  This seems to assume that we can perform `stat()`.  I suppose we can do a DAC override and just hit everything as UID `0` (downside is this causes errors in recent Fedoras and would not work over NFS in root squash mode).
2.  It does have the side-effect of doubling the number of metadata ops.

**Comment by abh3** [comment]
Hmmm, let's go on a different track. Is it not the case that you need to have switched to the correct uid before you even attempt to compute a checksum?

**Comment by bbockelm** [comment]
Correct!

@juztas - are these files world-readable by any chance (or, at least, readable by the `xrootd` user?)  If so, it's not clear that's a safe assumption going forward.

**Comment by juztas** [comment]
Due to another issue in xrootd-multiuser lib, permissions are 644 now, but the plan is to have 640 and not allow xrootd user to read.

**Comment by bbockelm** [comment]
@juztas: Gotcha - so, in all likelihood, we'll need this information in the `XrdCksManOss` class as well to ensure the file is opened with the correct permissions.

Also implies that we can't do the double-stat trick as that would be `xrootd-multiuser`-specific information that has to be leaked into the `XrdCksManOss` class.  :(

**Comment by abh3** [comment]
I'd like to point out that the SecEntity object is passed, if indirectly, to the checksum manager for get/calc if it is a proxy. The reason is that the proxy needs the entity information if it is actually proxying the client. See XrdOfs.cc:1423.

**Comment by bbockelm** [comment]
Not sure I'm following here - how does having the `tident` allow me to get back to the `SecEntity`?  Might need a few more breadcrumbs here...

**Comment by xrootd-dev** [comment]
No, look more closely; you can see that it passes an XrdOucEnv object 
which has a pointer to the SecEntity object.

**Comment by bbockelm** [comment]
Ah - was looking at an older branch!  I see it now, makes sense.

**Comment by bbockelm** [comment]
@abh3 - is there a reason why it's only done for the proxy?  Can it be done unconditionally?

Still not clear how to get this down to the `xattrlib` without doing various hacks.  I do see how this would provide a fix for the pieces in `XrdCksManOss`.

**Comment by abh3** [comment]
Not clear why it was restricted to the proxy. Likely because of ABI 
compatability. I will research this to see what happens if we just set it 
that way. It would mostly affect DPM as it's the only system that I know 
of that replaces that plugin.

**Comment by abh3** [comment]
Oh yes, could you do me a favor and use the new post authentication plugin,

https://xrootd.slac.stanford.edu/doc/dev50/sec_config.htm#_Toc42289896

to set the uid/gid only once irrespective of what authentication mechanism was used? This would centralize the nagging uid/gid lookups.

**Comment by bbockelm** [comment]
@abh3 - any updates as to whether we could have this information exported in non-proxy mode?

**Comment by xrootd-dev** [comment]
Sorry Brian, I haven't had the time to look at this closely. I will do so 
this coming week as I will finish up a big project I am working on.

Andy

**Comment by abh3** [comment]
It looks like it is safe to simply set that field to point to the env. However, I am still following up with the DPM people to make sure; though looking at the latest DPM source it would appear to be safe for DPM as well but I want confirmation.

**Comment by abh3** [comment]
OK, confirmed that it's OK to use hat field for every type of server. I will change that.

**Comment by bbockelm** [comment]
Hi @abh3 -

I don't see how this could work.  Consider this stack:

```
(gdb) bt
#0  XrdCksManOss::Calc (this=0x2416db0, Lfn=0x7f5c7c00b0a0 "/tmp/xrootd/fooGoogle", Cks=..., 
    doSet=1) at /home/cse496/bbockelm/projects/xrootd/src/XrdCks/XrdCksManOss.cc:101
#1  0x00007f5cbae9221e in XrdOfs::chksum (
    this=0x7f5cbb133a80 <XrdSfsGetDefaultFileSystem(XrdSfsFileSystem*, XrdSysLogger*, char const*, XrdOucEnv*)::XrdDefaultOfsFS>, Func=XrdSfsFileSystem::csCalc, 
    csName=0x7f5c7c00b2d0 "md5", Path=0x7f5c7c00b0a0 "/tmp/xrootd/fooGoogle", einfo=..., 
    client=0x0, opaque=0x0) at /home/cse496/bbockelm/projects/xrootd/src/XrdOfs/XrdOfs.cc:1568
#2  0x00007f5cb1bf43be in MultiuserFileSystem::chksum (this=0x23ba680, 
    Func=XrdSfsFileSystem::csCalc, csName=0x7f5c7c00b2d0 "md5", 
    path=0x7f5c7c00b0a0 "/tmp/xrootd/fooGoogle", eInfo=..., client=0x0, opaque=0x0)
    at /home/cse496/bbockelm/projects/xrootd-multiuser/src/multiuser.cpp:566
#3  0x00007f5cbae6ede0 in XrdXrootdProtocol::CheckSum (Stream=0x7f5c7c00c010, 
    argv=0x7f5caa2389e0, argc=4)
    at /home/cse496/bbockelm/projects/xrootd/src/XrdXrootd/XrdXrootdProtocol.cc:890
#4  0x00007f5cbab48cf7 in XrdOucProg::Run (this=0x23b9e50, Sp=0x7f5c7c00c010, 
    argV=0x7f5caa238c70, argC=3, envV=0x0)
    at /home/cse496/bbockelm/projects/xrootd/src/XrdOuc/XrdOucProg.cc:140
#5  0x00007f5cbab48f20 in XrdOucProg::Run (this=0x23b9e50, Sp=0x7f5c7c00c010, 
    arg1=0x7f5c7c00b2d0 "md5", arg2=0x7f5c7c00b0a0 "/tmp/xrootd/fooGoogle", 
    arg3=0x7f5c7c00bb80 "bbockelm.8:34@localhost", arg4=0x0)
    at /home/cse496/bbockelm/projects/xrootd/src/XrdOuc/XrdOucProg.cc:173
#6  0x00007f5cbae64a5e in XrdXrootdJob2Do::DoIt (this=0x7f5c7c00bf60)
    at /home/cse496/bbockelm/projects/xrootd/src/XrdXrootd/XrdXrootdJob.cc:168
#7  0x00007f5cbab88bf0 in XrdScheduler::Run (this=0x61bde0 <XrdGlobal::Sched>)
    at /home/cse496/bbockelm/projects/xrootd/src/Xrd/XrdScheduler.cc:382
#8  0x00007f5cbab87d24 in XrdStartWorking (carg=0x61bde0 <XrdGlobal::Sched>)
    at /home/cse496/bbockelm/projects/xrootd/src/Xrd/XrdScheduler.cc:88
#9  0x00007f5cbab2e886 in XrdSysThread_Xeq (myargs=0x7f5c9c001470)
    at /home/cse496/bbockelm/projects/xrootd/src/XrdSys/XrdSysPthread.cc:86
#10 0x00007f5cba6abdd5 in start_thread (arg=0x7f5caa239700) at pthread_create.c:307
#11 0x00007f5cb99adead in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

The checksum is run as a standalone `XrdXrootdJob` and only gets 4 parameters in addition to the `XrdOucStream` object.  The function `XrdXrootdProtocol::CheckSum` does not have access to the `XrdSecEntity` associated with the operation.  Hence, by time `XrdOfs::chksum` is invoked, there's no `XrdSecEntity` there either.  We can now pass down the pointer through the interface you changed - but it's still `nullptr`.

I looked into what it'd take to make the `XrdXrootdJob` infrastructure credential-aware but it's not clear what lifetime guarantees need to be made.

Can you take a look at this please?

There's still a small amount of other changes needed to pass the pointer down the stack; I'll submit those in a bit.  It's just that, again, we're currently passing down a `nullptr`.

**Comment by xrootd-dev** [comment]
Ah, you are using an old release. The R4 series only sets this for proxy 
servers. The R5 release sets it for everyone.

Andy

**Comment by juztas** [comment]
@bbockelm Maybe you had a chance to look into this?

**Comment by abh3** [comment]
Actually, this is rather unresolved. We don't have a good way of uniformly passing this when the uid/gid tracking plugin is used. We are still looking at what we can do about t but the fix won't be particularly easy.

**Comment by juztas** [comment]
@abh3 Can we re-open the issue?

**Comment by abh3** [comment]
I suppose we can reopen it but the only way this is going to get fixed is by reworking the uid/gid plugin. We are working on that approach as the way the original plugin was written is in conflict with the server's internal structure.

**Comment by abh3** [comment]
Derek and I have looked at the multi0user plugin. It indeed sets the uid-gid before calling the checksum routine so that whole routine, including extended attribute processing runs with the preset uid-gid. So, as far as we can tell it should work just fine. So, could you provide log output and some reasonable debug level that captures it not working? We seem to be missing that information.

**Comment by juztas** [comment]
How can I set a max of the maximum debug level for all xrootd and it's modules?

**Comment by xrootd-dev** [comment]
Simply start the daemon with the '-d' option.

**Comment by juztas** [comment]
Here:
```
-bash-4.2$ gfal-ls -l gsiftp://transfer.ultralight.org//mnt/hadoop/store/user/jbalcas/for-xrootd-devs/
-rw-rw-r--   1 0     0       1399379 Feb 22 07:15 strace.out
-rw-rw-r--   1 0     0         27765 Feb 22 07:15 xrootd.log
-bash-4.2$ gfal-ls -l root://xrootd.ultralight.org//store/user/jbalcas/for-xrootd-devs/
-r--r--r--   0 0     0       1399379 Feb 22 07:15 /strace.out
-r--r--r--   0 0     0         27765 Feb 22 07:15 /xrootd.log
```
p.s. you will need a cms proxy to access these files.

**Comment by abh3** [comment]
There is no way I can get a cms proxy cert. Could you put them somewhere I can simply download them? Send me the URLs in private email, if so.

**Comment by abh3** [comment]
Thank you Justas for the logs. This was sort of a tortuous journey with many misunderstanding on my part. Now that I've seen the logs, dissected the multi-user plugin, and understood the non-obvious interactions between that plugin and the rest of the plugins I understand the issue and it's not particularly easy to fix as there are several conflicting assumptions made by the two most relevant plugins: multi-user and the OFS plugin on two major levels. A lot has to do with the positioning of the multi-user plugin and it would be somewhat easier to solve if it fronted the oss plugin as opposed to the sfs plugin. That said, my conversation with Derek indicated that it would be too disruptive to change the architecture at this point, so we won't take that route.

Problem #1: As indicated by Brian (and missed the fine point at the time) the failure occurs because of the way checksums are computed in the background. Once we establish that a checksum is missing a background job is run to compute it. However, since the server has already verified the permissions to do this, it runs the computation without involving authorization as it's not needed and may fail if it tried to do so in background without the multi-user plugin. This is done by not passing the client information to the checksum call. Of course, this conflicts with  the multi-user plugin which does require client information. The server really has no way of knowing that; so the solution here will need some changes in the checksum launch code as well as the multi-user plugin to keep compatibility.

Problem #2: The file /storage/cms/store/temp/jbalcas/test-xrootd14 was open using the POSC option. Since the uid/gix was set prior to calling the underlying plugin, when called the control files created to track the file so it could be deleted were written using the set uid/gid not the xroot uid/gid. When the file was finally closed using the xroot user we got a failure:
Unable to fchmod /storage/cms/store/temp/jbalcas/test-xrootd14; operation not permitted

This likely means that the control files were not cleaned up. It also means that the daemon will not be able to restart should the control files indicate that the file still exists because it will not be able to delete the file. This is inherent to the multi-user plugin positioning and there really is no way to easily get around the problem. The short answer is that the multi-user plugin has to indicate that it does not support POSC because it really does not.

As I said, I can fix both problems with a specific patch for multi-user but I would need to change the multi-user plugin as well. There is no other way to do this the way it is written now. I will generate the appropriate pull request for the multi-user plugin to a) get the correct username for background checksums yet not apply authorization at that point and
b) indicate that POSC is not supported via the multi-user plugin (this also will turn off checkpointing).

Note that these fixes will apply to R5.1.0 as there is no easy way to backport them to the 5.x series. OK?
